### ISSUE TITLE
Reactor Netty Http client over .protocol(HttpProtocol.H2) is blocking

### ISSUE THREAD

**Issue by patpatpat123** [reactor/reactor-netty#1508]
Hello Reactor Netty Team,

I might have found a blocking call using the reactive http client provided by Netty.
For a Spring Webflux 2.4.2 project, we perform regression testing with blockhound enabled.

During our regression, blockhound barked with the following:

```
Caused by: reactor.blockhound.BlockingOperationError: Blocking call! java.io.FileInputStream#readBytes
	at java.base/java.io.FileInputStream.readBytes(FileInputStream.java) ~[na:na]
	at java.base/java.io.FileInputStream.read(FileInputStream.java:279) ~[na:na]
	at java.base/sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284) ~[na:na]
	at java.base/sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326) ~[na:na]
	at java.base/sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178) ~[na:na]
	at java.base/java.io.InputStreamReader.read(InputStreamReader.java:185) ~[na:na]
	at java.base/java.io.BufferedReader.fill(BufferedReader.java:161) ~[na:na]
	at java.base/java.io.BufferedReader.readLine(BufferedReader.java:326) ~[na:na]
	at java.base/java.io.BufferedReader.readLine(BufferedReader.java:392) ~[na:na]
	at io.netty.resolver.dns.UnixResolverDnsServerAddressStreamProvider.parseEtcResolverSearchDomains(UnixResolverDnsServerAddressStreamProvider.java:373) ~[netty-resolver-dns-4.1.58.Final.jar:4.1.58.Final]
	at io.netty.resolver.dns.UnixResolverDnsServerAddressStreamProvider.parseEtcResolverSearchDomains(UnixResolverDnsServerAddressStreamProvider.java:354) ~[netty-resolver-dns-4.1.58.Final.jar:4.1.58.Final]
	at io.netty.resolver.dns.DnsNameResolver.<clinit>(DnsNameResolver.java:137) ~[netty-resolver-dns-4.1.58.Final.jar:4.1.58.Final]
	at io.netty.resolver.dns.DnsNameResolverBuilder.<init>(DnsNameResolverBuilder.java:49) ~[netty-resolver-dns-4.1.58.Final.jar:4.1.58.Final]
	at reactor.netty.transport.NameResolverProvider.newNameResolverGroup(NameResolverProvider.java:405) ~[reactor-netty-core-1.0.3.jar:1.0.3]
	at reactor.netty.transport.ClientTransportConfig.getOrCreateDefaultResolver(ClientTransportConfig.java:230) ~[reactor-netty-core-1.0.3.jar:1.0.3]
	at reactor.netty.transport.ClientTransportConfig.resolverInternal(ClientTransportConfig.java:216) ~[reactor-netty-core-1.0.3.jar:1.0.3]
	at reactor.netty.http.client.HttpClientConfig.resolverInternal(HttpClientConfig.java:420) ~[reactor-netty-http-1.0.3.jar:1.0.3]
	at reactor.netty.http.client.HttpClientConnect$MonoHttpConnect.lambda$subscribe$0(HttpClientConnect.java:262) ~[reactor-netty-http-1.0.3.jar:1.0.3]
	at reactor.core.publisher.MonoCreate.subscribe(MonoCreate.java:57) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.FluxRetryWhen.subscribe(FluxRetryWhen.java:76) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.MonoRetryWhen.subscribeOrReturn(MonoRetryWhen.java:46) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:57) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.netty.http.client.HttpClientConnect$MonoHttpConnect.subscribe(HttpClientConnect.java:269) ~[reactor-netty-http-1.0.3.jar:1.0.3]
	at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.Mono.subscribe(Mono.java:4046) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.MonoZip.subscribe(MonoZip.java:128) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.MonoCacheTime.subscribeOrReturn(MonoCacheTime.java:134) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.Mono.subscribe(Mono.java:4031) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.MonoZip.subscribe(MonoZip.java:128) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.Mono.subscribe(Mono.java:4046) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.MonoZip.subscribe(MonoZip.java:128) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.Flux.subscribe(Flux.java:8147) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.FluxUsingWhen.subscribe(FluxUsingWhen.java:93) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.MonoCacheTime.subscribeOrReturn(MonoCacheTime.java:134) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.Mono.subscribe(Mono.java:4031) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.MonoZip.subscribe(MonoZip.java:128) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:157) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1789) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.MonoFlatMap$FlatMapInner.onNext(MonoFlatMap.java:249) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:79) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.FluxPeekFuseable$PeekFuseableSubscriber.onNext(FluxPeekFuseable.java:210) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.FluxPeekFuseable$PeekFuseableSubscriber.onNext(FluxPeekFuseable.java:210) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1789) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.MonoIgnoreThen$ThenAcceptInner.onNext(MonoIgnoreThen.java:305) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1789) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:151) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1789) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.MonoZip$ZipCoordinator.signal(MonoZip.java:251) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.MonoZip$ZipInner.onNext(MonoZip.java:336) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.MonoPeekTerminal$MonoTerminalPeekSubscriber.onNext(MonoPeekTerminal.java:180) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.FluxDefaultIfEmpty$DefaultIfEmptySubscriber.onNext(FluxDefaultIfEmpty.java:100) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.FluxPeek$PeekSubscriber.onNext(FluxPeek.java:199) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.FluxSwitchIfEmpty$SwitchIfEmptySubscriber.onNext(FluxSwitchIfEmpty.java:73) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:79) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1789) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:151) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.onNext(FluxContextWrite.java:107) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.FluxMapFuseable$MapFuseableConditionalSubscriber.onNext(FluxMapFuseable.java:295) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.FluxFilterFuseable$FilterFuseableConditionalSubscriber.onNext(FluxFilterFuseable.java:337) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1789) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.MonoCollect$CollectSubscriber.onComplete(MonoCollect.java:159) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.FluxMap$MapSubscriber.onComplete(FluxMap.java:142) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.FluxPeek$PeekSubscriber.onComplete(FluxPeek.java:259) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.FluxMap$MapSubscriber.onComplete(FluxMap.java:142) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.netty.channel.FluxReceive.terminateReceiver(FluxReceive.java:457) ~[reactor-netty-core-1.0.3.jar:1.0.3]
	at reactor.netty.channel.FluxReceive.drainReceiver(FluxReceive.java:258) ~[reactor-netty-core-1.0.3.jar:1.0.3]
	at reactor.netty.channel.FluxReceive.request(FluxReceive.java:130) ~[reactor-netty-core-1.0.3.jar:1.0.3]
	at reactor.core.publisher.FluxMap$MapSubscriber.request(FluxMap.java:162) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.FluxPeek$PeekSubscriber.request(FluxPeek.java:137) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.FluxMap$MapSubscriber.request(FluxMap.java:162) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.MonoCollect$CollectSubscriber.onSubscribe(MonoCollect.java:103) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.FluxMap$MapSubscriber.onSubscribe(FluxMap.java:92) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.FluxPeek$PeekSubscriber.onSubscribe(FluxPeek.java:170) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.core.publisher.FluxMap$MapSubscriber.onSubscribe(FluxMap.java:92) ~[reactor-core-3.4.2.jar:3.4.2]
	at reactor.netty.channel.FluxReceive.startReceiver(FluxReceive.java:168) ~[reactor-netty-core-1.0.3.jar:1.0.3]
	at reactor.netty.channel.FluxReceive.lambda$subscribe$2(FluxReceive.java:147) ~[reactor-netty-core-1.0.3.jar:1.0.3]
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164) ~[netty-common-4.1.58.Final.jar:4.1.58.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472) ~[netty-common-4.1.58.Final.jar:4.1.58.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500) ~[netty-transport-4.1.58.Final.jar:4.1.58.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) ~[netty-common-4.1.58.Final.jar:4.1.58.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.58.Final.jar:4.1.58.Final]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.58.Final.jar:4.1.58.Final]
	at java.base/java.lang.Thread.run(Thread.java:834) ~[na:na]

``` 

Therefore, we started to investigate (and investigating blocking call is not easy, but that is a topic for another day 😛)

We tried putting breakpoints here and there, and we believe to have found the root cause.
In our code, we have a piece of code:

```
    @Bean
    public ReactiveRegistrationClient registrationClient(ClientProperties clientProperties) {
        final Builder webClient = getWebClient().mutate().clientConnector(new ReactorClientHttpConnector(HttpClient.create().protocol(HttpProtocol.H2).wiretap(true)));
        webClient.filter(ExchangeFilterFunctions.basicAuthentication(Objects.requireNonNull(clientProperties.getUsername()), Objects.requireNonNull(clientProperties.getPassword())));
        return new ReactiveRegistrationClient(webClient.build(), clientProperties.getReadTimeout());
    }
```

we then removed the ```.protocol(HttpProtocol.H2)```
```
    @Bean
    public ReactiveRegistrationClient registrationClient(ClientProperties clientProperties) {
        final Builder webClient = getWebClient().mutate().clientConnector(new ReactorClientHttpConnector(HttpClient.create().wiretap(true)));
        webClient.filter(ExchangeFilterFunctions.basicAuthentication(Objects.requireNonNull(clientProperties.getUsername()), Objects.requireNonNull(clientProperties.getPassword())));
        return new ReactiveRegistrationClient(webClient.build(), clientProperties.getReadTimeout());
    }
```

And the blocking call disappeared!!!

My initial conclusion, is that ```.protocol(HttpProtocol.H2)``` is the root cause of the blocking call.

Hence, would it be possible ```.protocol(HttpProtocol.H2)``` is blocking in a reactive world please?
Or please let me know what am I not doing correctly.

Thank you for your help.

**Comment by violetagg** [comment]
@patpatpat123 I've already proposed a change to Netty https://github.com/netty/netty/pull/11009
As a workaround you can add your own BlockHound integration see the issue https://github.com/netty/netty/issues/11004 

Closing this issue and marking `for/netty`

**Comment by violetagg** [comment]
@patpatpat123 Another workaround would be to use the new functionality for `warmup` that will load the DNS resolver before the first request. https://projectreactor.io/docs/netty/release/reference/index.html#_eager_initialization_4

**Comment by patpatpat123** [comment]
Indeed!

```
@Bean
    public ReactiveRegistrationClient registrationClient(ClientProperties clientProperties) {
        final HttpClient httpClient = HttpClient.create().protocol(HttpProtocol.H2).wiretap(true).secure(sslContextSpec -> sslContextSpec.sslContext(getSslContextBuilder()));
        httpClient.warmup().block();
        final Builder webClient = getWebClient().mutate().clientConnector(new ReactorClientHttpConnector(httpClient));
        webClient.filter(ExchangeFilterFunctions.basicAuthentication(Objects.requireNonNull(clientProperties.getUsername()), Objects.requireNonNull(clientProperties.getPassword())));
        return new ReactiveRegistrationClient(webClient.build(), clientProperties.getReadTimeout());
    }
```

Is not blocking anymore! (although a bit scary to now see ```.block()``` is the code 😛

The fix you have provided here https://github.com/netty/netty/pull/11009 is much cleaner.
We are looking forward to seeing this new version!

Thank you for the help provided

**Comment by violetagg** [comment]
> Indeed!
> 
> ```
> @Bean
>     public ReactiveRegistrationClient registrationClient(ClientProperties clientProperties) {
>         final HttpClient httpClient = HttpClient.create().protocol(HttpProtocol.H2).wiretap(true).secure(sslContextSpec -> sslContextSpec.sslContext(getSslContextBuilder()));
>         httpClient.warmup().block();
>         final Builder webClient = getWebClient().mutate().clientConnector(new ReactorClientHttpConnector(httpClient));
>         webClient.filter(ExchangeFilterFunctions.basicAuthentication(Objects.requireNonNull(clientProperties.getUsername()), Objects.requireNonNull(clientProperties.getPassword())));
>         return new ReactiveRegistrationClient(webClient.build(), clientProperties.getReadTimeout());
>     }
> ```
> 
> Is not blocking anymore! (although a bit scary to now see `.block()` is the code 😛

I would recommend to use always `block` with a proper timeout.
> 
> The fix you have provided here [netty/netty#11009](https://github.com/netty/netty/pull/11009) is much cleaner.
> We are looking forward to seeing this new version!
> 

The fix in Netty does not remove the blocking call bu t allows a blocking call inside thread marked as non-blocking.

> Thank you for the help provided
